src/util/threadpool.rs

```diff
@@ -0,0 +1,609 @@
+}
+
+pub struct PushError<T>(pub T);
+
+// `BigGroupThrottledQueue` tries to throttle group's concurrency to
+//  `group_concurrency_on_busy` when it's busy.
+// When one worker asks a task to run, it schedules in the following way:
+// 1. Find out which group has a running number that is smaller than
+//    that of `group_concurrency_on_busy`.
+// 2. If more than one group meets the first point, run the one who
+//    comes first.
+// If no group meets the first point, choose according to the following rules:
+// 1. Select the groups with least running tasks.
+// 2. If more than one group meets the first point,choose the one
+//     whose task comes first(with the minimum task's ID)
+pub struct BigGroupThrottledQueue<T> {
+    // Tasks in pending. The Tasks in `pending_tasks` have higher priority than
+    // tasks in `waiting_queue`.
+    pending_tasks: BinaryHeap<Task<T>>,
+    // group_id => tasks array. If `group_concurrency[group_id]` is bigger than
+    // `group_concurrency_on_busy`(which means the number of on-going tasks is
+    // more than `group_concurrency_on_busy`), the rest of the group's tasks
+    // would be pushed into `waiting_queue[group_id]`
+    waiting_queue: BTreeMap<T, VecDeque<Task<T>>>,
+    // group_id => running_num+pending_num(in `pending_tasks`). It means at most
+    // `group_concurrency[group_id]` tasks of the group may be running
+    group_concurrency: HashMap<T, usize>,
+    // The maximum number of threads that each group can run when the pool is busy.
+    // Each value in `group_concurrency` shouldn't be bigger than this value.
+    group_concurrency_on_busy: usize,
+}
+
+
+impl<T: Hash + Eq + Ord + Send + Clone> BigGroupThrottledQueue<T> {
+    pub fn new(group_concurrency_on_busy: usize) -> BigGroupThrottledQueue<T> {
+        BigGroupThrottledQueue {
+            group_concurrency: HashMap::new(),
+            waiting_queue: BTreeMap::new(),
+            pending_tasks: BinaryHeap::new(),
+            group_concurrency_on_busy: group_concurrency_on_busy,
+        }
+    }
+
+    // Try push into pending. Return none on success,return PushError(task) on failed.
+    #[inline]
+    fn try_push_into_pending(&mut self, task: Task<T>) -> Result<(), PushError<Task<T>>> {
+        let count = self.group_concurrency.entry(task.group_id.clone()).or_insert(0);
+        if *count >= self.group_concurrency_on_busy {
+            return Err(PushError(task));
+        }
+        *count += 1;
+        self.pending_tasks.push(task);
+        Ok(())
+    }
+
+    #[inline]
+    fn pop_from_waiting_queue(&mut self) -> Option<Task<T>> {
+        let group_id = {
+            // Groups in waiting_queue wouldn't too much, so iterate the map
+            // is quick.
+            let best_group = self.waiting_queue
+                .iter()
+                .map(|(group_id, waiting_queue)| {
+                    (self.group_concurrency[group_id], waiting_queue[0].id, group_id)
+                })
+                .min();
```

Comment:
FYI, it could use data structure like SkipList/Heap which is fast for insertion/deletion/ordered access to track the relationship of (lowest concurrency, low id) -> group_id, to avoid the iteration.